crates/hyperdrive-math/src/lp/remove.rs

```diff
@@ -1,6 +1,312 @@
+
+        let mut state = self.clone();
+        // Burn the LP's shares.
+        state.info.lp_total_supply -= lp_shares.into();
+        let active_lp_total_supply = active_lp_total_supply - lp_shares;
+
+        // Mint an equivalent amount of withdrawal shares.
+        let withdrawal_shares_total_supply = withdrawal_shares_total_supply + lp_shares;
+
+        // Redeem as many of the withdrawal shares as possible.
+        let (proceeds, withdrawal_shares_redeemed) = state.calculate_redeem_withdrawal_shares(
+            current_block_timestamp,
+            active_lp_total_supply,
+            withdrawal_shares_total_supply,
+            lp_shares,
+            vault_share_price,
+            min_output_per_share,
+        )?;
+        let withdrawal_shares = lp_shares - withdrawal_shares_redeemed;
+
+        Ok((proceeds, withdrawal_shares))
+    }
+
+    /// Redeems withdrawal shares by giving the LP a pro-rata amount of the
+    /// withdrawal pool's proceeds. This function redeems the maximum amount of
+    /// the specified withdrawal shares given the amount of withdrawal shares
+    /// ready to withdraw.
+    pub fn calculate_redeem_withdrawal_shares(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint,
+        withdrawal_shares_total_supply: FixedPoint,
+        withdrawal_shares: FixedPoint,
+        vault_share_price: FixedPoint,
+        min_output_per_share: FixedPoint,
+    ) -> Result<(FixedPoint, FixedPoint)> {
+        // Distribute the excess idle to the withdrawal pool. If the distribute
+        // excess idle calculation fails, we proceed with the calculation since
+        // LPs should be able to redeem their withdrawal shares for existing
+        // withdrawal proceeds regardless of whether or not idle could be
+        // distributed.
+        let (withdrawal_shares_redeemed, share_proceeds, _success) = self
+            .calculate_distribute_excess_idle_safe(
+                current_block_timestamp,
+                active_lp_total_supply,
+                withdrawal_shares_total_supply,
+                SHARE_PROCEEDS_MAX_ITERATIONS,
+            )?;
+
+        // Update withdrawal pool after distributing excess idle.
+        let ready_to_withdraw =
+            self.withdrawal_shares_ready_to_withdraw() + withdrawal_shares_redeemed;
+        let withdrawal_share_proceeds = self.withdrawal_shares_proceeds() + share_proceeds;
+
+        // Clamp the shares to the total amount of shares ready for withdrawal
+        // to avoid unnecessary reverts. We exit early if the user has no shares
+        // available to redeem.
+        let mut withdrawal_shares_redeemed = withdrawal_shares;
+        if withdrawal_shares_redeemed > ready_to_withdraw {
+            withdrawal_shares_redeemed = ready_to_withdraw;
+        }
+        if withdrawal_shares_redeemed == fixed!(0) {
+            return Ok((fixed!(0), fixed!(0)));
+        }
+
```

Comment:
solidity burns shares from the user here; was that skipped on purpose?

Reply:
skipped on purpose, rust doesn't have that state.  these are `calculate_x` functions so as long as we return a correct calculation we can ignore resulting state updates.  Also, by state I mean contract state, i.e. things like user token balances, not the `State` object we pass around.

Reply:
ok cool thanks for clarifying :)